solutions/platinum/cses-1742.mdx

```diff
@@ -0,0 +1,270 @@
+
+		if (seg.a > seg.x) swap(seg.a, seg.x);
+		if (seg.b > seg.y) swap(seg.b, seg.y);
+		max_height = max(max_height, seg.b);
+		max_height = max(max_height, seg.y);
+	}
+}
+
+bool intersect(int idx) {
+	vector<event> events;
+	for (int i = 0; i <= idx; i++) {
+		if (segments[i].a == segments[i].x) {
+			events.push_back({segments[i].a, segments[i].b, i});
+		} else {
+			events.push_back({segments[i].a, segments[i].b, BEGINNING_VALUE});
+			events.push_back({segments[i].x, segments[i].y, ENDING_VALUE});
+		}
+	}
+	sort(all(events));
+
+	Segtree sg(max_height);
+	Segment last_vertical_line = {0, 0, 0, 0};
+	for (auto event : events) {
+		if (event.id == BEGINNING_VALUE) {
+			// Beginning of a horizontal line
+			sg.update(1, 1, max_height, event.y, 1);
+			// Check for horizontal-horizontal intersection
+			if (sg.query(1, 1, max_height, event.y, event.y) > 1) return true;
+		} else if (event.id == ENDING_VALUE) {
+			// Ending of a horizontal line
+			sg.update(1, 1, max_height, event.y, -1);
+		} else {
+			// Checking for vertical-horizontal intersection
+			if (sg.query(1, 1, max_height, segments[event.id].b,
+			             segments[event.id].y) > 0)
+				return true;
+			// Checking for vertical-vertical intersection
+			if (segments[event.id].a == last_vertical_line.a &&
+			    segments[event.id].b <= last_vertical_line.y &&
+			    segments[event.id].y > last_vertical_line.b)
+				return true;
+			last_vertical_line = segments[event.id];
+		}
+	}
+	return false;
+}
+
+int main() {
+	cin >> n;
+
+	dir['R'] = {1, 0};
+	dir['L'] = {-1, 0};
+	dir['U'] = {0, 1};
+	dir['D'] = {0, -1};
+
+	ll a = 0, b = 0;
+	for (int i = 0; i < n; i++) {
+		char c;
+		int steps;
+		cin >> c >> steps;
+
+		ll x = a + dir[c].first * steps;
+		ll y = b + dir[c].second * steps;
+
+		if (i > 0) {
+			a += dir[c].first;
+			b += dir[c].second;
+		}
+
+		direction.push_back(c);
+
+		segments.push_back({a, b, x, y});
+		original_segments.push_back({a, b, x, y});
+
+		a = x, b = y;
+	}
+
+	coordinate_compression(segments);
+
+	// Binary search
+	int l = 0, r = n - 1, last = r;
+	while (l <= r) {
+		// Check if the first mid segments intersect
+		int mid = (l + r) >> 1;
+		if (!intersect(mid)) l = mid + 1, last = mid;
+		else r = mid - 1;
+	}
+
+	auto dist = [](Segment seg) {
+		return abs(seg.x - seg.a) + abs(seg.y - seg.b) + 1;
+	};
+	// Check if segment a and b intersect
+	auto intersect = [](Segment a, Segment b) {
+		if (a.a > a.x) swap(a.a, a.x);
+		if (a.b > a.y) swap(a.b, a.y);
+		if (b.b > b.y) swap(b.b, b.y);
+		if (b.a > b.x) swap(b.a, b.x);
+		if (a.b == a.y && b.b == b.y && a.b == b.b)
+			return a.a <= b.x && a.x > b.a;
+		if (a.a == a.x && b.a == b.x && a.a == b.a)
+			return a.b <= b.y && a.y > b.b;
+		if (a.a == a.x && b.b == b.y)
+			return a.b <= b.b && b.b <= a.y && b.a <= a.a && b.x >= a.a;
+		if (a.b == a.y && b.a == b.x)
+			return b.b <= a.b && a.b <= b.y && a.a <= b.a && a.x >= b.a;
+		return false;
+	};
```

Comment:
Is there a better way to check if they intersect without writing 15 lines of code?